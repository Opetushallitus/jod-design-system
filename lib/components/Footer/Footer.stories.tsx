import type { Meta, StoryObj } from '@storybook/react';

import { Footer, type FooterProps } from './Footer';

const meta = {
  title: 'Footer',
  component: Footer,
  tags: ['autodocs'],
} satisfies Meta<typeof Footer>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  layout: 'fullscreen',
  viewport: {
    defaultViewport: 'desktop',
  },
};

const items: FooterProps['items'] = [
  {
    component: ({ className }) => (
      <a href="/about" className={className}>
        Tietoa palvelusta ja käyttöohjeet
      </a>
    ),
  },
  {
    component: ({ className }) => (
      <a href="/cookies" className={className}>
        Evästekäytäntö
      </a>
    ),
  },
  {
    component: ({ className }) => (
      <a href="/sources" className={className}>
        Datalähteet
      </a>
    ),
  },
  {
    component: ({ className }) => (
      <a href="/terms" className={className}>
        Käyttöehdot
      </a>
    ),
  },
  {
    component: ({ className }) => (
      <a href="/accessibility" className={className}>
        Saavutettavuusseloste
      </a>
    ),
  },
  {
    component: ({ className }) => (
      <a href="/privacy" className={className}>
        Tietosuojaseloste
      </a>
    ),
  },
];

const logos: FooterProps['logos'] = [
  {
    component: ({ key, className }) => (
      <a key={key} href="/logo1" className={className}>
        Logo 1
      </a>
    ),
  },
  {
    component: ({ key, className }) => (
      <a key={key} href="/logo2" className={className}>
        Logo 2
      </a>
    ),
  },
  {
    component: ({ key, className }) => (
      <a key={key} href="/logo3" className={className}>
        Logo 3
      </a>
    ),
  },
];

const copyright = '© JOD 2024. Kaikki oikeudet pidätetään.';

export const Light: Story = {
  parameters: {
    ...parameters,
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A8163',
    },
    docs: {
      description: {
        story: 'This is a light footer component with navigation items.',
      },
    },
  },
  args: {
    items,
    logos,
    copyright,
  },
};

export const Dark: Story = {
  parameters: {
    ...parameters,
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A8189',
    },
    docs: {
      description: {
        story: 'This is a dark footer component with navigation items.',
      },
    },
    backgrounds: {
      default: 'jod-black',
      values: [{ name: 'jod-black', value: '#000000' }],
    },
  },
  args: {
    items,
    logos,
    copyright,
    variant: 'dark',
  },
};

export const MobileDark: Story = {
  parameters: {
    ...parameters,
    viewport: {
      defaultViewport: 'mobile',
    },
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=746%3A44679',
    },
    docs: {
      description: {
        story: 'This is a mobie footer component with navigation items.',
      },
    },
    backgrounds: {
      default: 'jod-black',
      values: [{ name: 'jod-black', value: '#000000' }],
    },
  },
  args: {
    items,
    logos,
    copyright,
    variant: 'dark',
  },
};
