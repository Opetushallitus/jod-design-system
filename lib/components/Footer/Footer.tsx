import { forwardRef } from 'react';

export interface FooterProps {
  /** Navigation items */
  items?: {
    component: React.ComponentType<{ className: string }>;
  }[];
  /** Collection of logos */
  logos?: {
    component: React.ComponentType<{ key: number; className: string }>;
  }[];
  /** Copyright text */
  copyright?: string;
  /** Variant of the footer */
  variant?: 'light' | 'dark';
}

/**
 * This component is a footer that displays navigation items, logos, and a copyright.
 */
export const Footer = forwardRef<HTMLDivElement, FooterProps>(function Footer(
  { items, logos, copyright, variant = 'light' }: FooterProps,
  ref,
) {
  return (
    <footer
      ref={ref}
      className={`py-9 text-button-md sm:px-6 print:hidden ${variant === 'light' ? 'bg-white text-black' : 'bg-black text-white'}`}
    >
      <div className="mx-auto flex flex-col gap-9 sm:max-w-[1090px]">
        {items && (
          <ul className="px-5 sm:px-4">
            {items?.map((item, index) => (
              <li key={index} className="px-2 py-[6px]">
                <item.component className="hover:underline" />
              </li>
            ))}
          </ul>
        )}
        {logos && (
          <div className="flex flex-col gap-6 px-5 sm:flex-row sm:px-0">
            {logos.map((logo, index) => (
              <logo.component
                key={index}
                className={`flex h-10 w-full items-center justify-center border-2 border-solid outline-none transition ${
                  variant === 'light'
                    ? 'border-black text-black hover:bg-black hover:text-white focus:bg-black focus:text-white'
                    : 'border-white text-white hover:bg-white hover:text-black focus:bg-white focus:text-black'
                }`}
              />
            ))}
          </div>
        )}
        {copyright && <div className="mt-9 flex justify-end px-5 sm:px-0">{copyright}</div>}
      </div>
    </footer>
  );
});
