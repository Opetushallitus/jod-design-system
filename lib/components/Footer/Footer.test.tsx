import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { Footer } from './Footer';

const mockOkmLabel = 'Opetus- ja kulttuuriministeriö';
const mockTemLabel = 'Työ- ja elinkeinoministeriö';
const mockOphLabel = 'Opetushallitus';
const mockKehaLabel = 'KEHA-keskus';
const mockCooperationTitle = 'Osaamipolku on toteutettu yhteistyössä seuraavien tahojen toimesta.';
const mockFundingTitle = 'Palvelu on rahoitettu Euroopan RFF-rahoituksella.';
const mockCopyright = '© Osaamispolku 2025. Kaikki oikeudet pidätetään.';

const mockMoreInfoLinks = [
  {
    href: '/fi/tietoa-palvelusta',
    label: 'Tietoa palvelusta',
  },
  {
    href: '/fi/tietosuojaseloste-ja-evasteet',
    label: 'Tietosuojaselosteet ja evästeet',
  },
  {
    href: '/fi/datalahteet',
    label: 'Datalähteet',
  },
  {
    href: '/fi/tietoa-tekoalyn-kaytosta',
    label: 'Tietoa tekoälyn käytöstä',
  },
  {
    href: '/fi/tietoa-saavutettavuudesta',
    label: 'Saavutettavuusseloste',
  },
];

const mockMoreInfoTitle = 'Haluatko tietää lisää Osaamispolusta?';
const mockMoreInfoDescription =
  'Mietityttääkö tietosuoja tai tekoälyn hyödyntäminen palvelussa? Alta löydät kootusti yleistä tietoa palvelusta ja sen käytöstä.';

const mockFeedbackTitle = '';
const mockFeedbackContent = '';
const mockFeedbackButtonLabel = '';
const mockFeedbackBgImageClassName = '';

const FooterComponent = (
  <Footer
    language="fi"
    okmLabel={mockOkmLabel}
    temLabel={mockTemLabel}
    ophLabel={mockOphLabel}
    kehaLabel={mockKehaLabel}
    cooperationTitle={mockCooperationTitle}
    fundingTitle={mockFundingTitle}
    moreInfoLinks={mockMoreInfoLinks}
    moreInfoTitle={mockMoreInfoTitle}
    moreInfoDescription={mockMoreInfoDescription}
    copyright={mockCopyright}
    feedbackTitle={mockFeedbackTitle}
    feedbackContent={mockFeedbackContent}
    feedbackButtonLabel={mockFeedbackButtonLabel}
    feedbackOnClick={vi.fn()}
    feedbackBgImageClassName={mockFeedbackBgImageClassName}
    externalLinkIconAriaLabel="Linkki johtaa palvelun ulkopuolelle"
  />
);

describe('Footer', () => {
  it('renders', () => {
    const { container } = render(FooterComponent);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders copyright', () => {
    const { container } = render(FooterComponent);
    const copyright = screen.getByText('© Osaamispolku 2025. Kaikki oikeudet pidätetään.');
    expect(copyright).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('emits data-testid when provided', () => {
  render(
    <Footer
      language="fi"
      okmLabel={mockOkmLabel}
      temLabel={mockTemLabel}
      ophLabel={mockOphLabel}
      kehaLabel={mockKehaLabel}
      cooperationTitle={mockCooperationTitle}
      fundingTitle={mockFundingTitle}
      moreInfoLinks={mockMoreInfoLinks}
      moreInfoTitle={mockMoreInfoTitle}
      moreInfoDescription={mockMoreInfoDescription}
      copyright={mockCopyright}
      feedbackTitle={mockFeedbackTitle}
      feedbackContent={mockFeedbackContent}
      feedbackButtonLabel={mockFeedbackButtonLabel}
      feedbackOnClick={vi.fn()}
      feedbackBgImageClassName={mockFeedbackBgImageClassName}
      externalLinkIconAriaLabel="Linkki johtaa palvelun ulkopuolelle"
      dataTestId="footer"
    />,
  );
  expect(screen.getByTestId('footer')).toBeInTheDocument();
});
