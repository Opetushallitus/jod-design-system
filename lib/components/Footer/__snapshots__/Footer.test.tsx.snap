// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > renders copyright 1`] = `
<footer
  class="py-9 text-button-md sm:px-6 print:hidden bg-white text-black"
>
  <div
    class="mx-auto flex flex-col gap-9 sm:max-w-[1090px]"
  >
    <div
      class="mt-9 flex justify-end px-5 sm:px-0"
    >
      © 2024
    </div>
  </div>
</footer>
`;

exports[`Footer > renders footer with dark variant 1`] = `
<footer
  class="py-9 text-button-md sm:px-6 print:hidden bg-black text-white"
>
  <div
    class="mx-auto flex flex-col gap-9 sm:max-w-[1090px]"
  >
    <ul
      class="px-5 sm:px-4"
    >
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 1
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 2
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 3
        </div>
      </li>
    </ul>
    <div
      class="flex flex-col gap-6 px-5 sm:flex-row sm:px-0"
    >
      <div>
        Logo 1
      </div>
      <div>
        Logo 2
      </div>
    </div>
    <div
      class="mt-9 flex justify-end px-5 sm:px-0"
    >
      © 2024
    </div>
  </div>
</footer>
`;

exports[`Footer > renders footer with light variant 1`] = `
<footer
  class="py-9 text-button-md sm:px-6 print:hidden bg-white text-black"
>
  <div
    class="mx-auto flex flex-col gap-9 sm:max-w-[1090px]"
  >
    <ul
      class="px-5 sm:px-4"
    >
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 1
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 2
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 3
        </div>
      </li>
    </ul>
    <div
      class="flex flex-col gap-6 px-5 sm:flex-row sm:px-0"
    >
      <div>
        Logo 1
      </div>
      <div>
        Logo 2
      </div>
    </div>
    <div
      class="mt-9 flex justify-end px-5 sm:px-0"
    >
      © 2024
    </div>
  </div>
</footer>
`;

exports[`Footer > renders items 1`] = `
<footer
  class="py-9 text-button-md sm:px-6 print:hidden bg-white text-black"
>
  <div
    class="mx-auto flex flex-col gap-9 sm:max-w-[1090px]"
  >
    <ul
      class="px-5 sm:px-4"
    >
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 1
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 2
        </div>
      </li>
      <li
        class="px-2 py-[6px]"
      >
        <div>
          Item 3
        </div>
      </li>
    </ul>
  </div>
</footer>
`;

exports[`Footer > renders logos 1`] = `
<footer
  class="py-9 text-button-md sm:px-6 print:hidden bg-white text-black"
>
  <div
    class="mx-auto flex flex-col gap-9 sm:max-w-[1090px]"
  >
    <div
      class="flex flex-col gap-6 px-5 sm:flex-row sm:px-0"
    >
      <div>
        Logo 1
      </div>
      <div>
        Logo 2
      </div>
    </div>
  </div>
</footer>
`;
